pkg/cache/v2/snapshot.go

```diff
@@ -194,3 +218,44 @@ func (s *Snapshot) GetVersion(typeURL string) string {
 	}
 	return s.Resources[typ].Version
 }
+
+// GetVersionMap will construct a verison map off the current state of a snapshot
+func (s *Snapshot) GetVersionMap() (map[string]map[string]string, error) {
+	if s == nil {
+		return nil, fmt.Errorf("missing snapshot")
+	}
+
+	// We need to initialize the map before we process anything since it is nested
+	versionMap := initializeVMap()
+	for i := 0; i < int(types.UnknownType); i++ {
+		typeURL := GetResponseTypeURL(types.ResponseType(i))
+		resources := s.GetResources(typeURL)
+
+		for _, r := range resources {
+			// hash our verison in here and build the version map
+			marshaledResource, err := MarshalResource(r)
+			if err != nil {
+				return nil, fmt.Errorf("failed to marshal resource: %v", err)
+			}
+			v := HashResource(marshaledResource)
+			if v != "" {
+				return nil, fmt.Errorf("failed to build resource version from hash: %v", err)
+			}
+
+			versionMap[typeURL][GetResourceName(r)] = v
+		}
+	}
+
+	return versionMap, nil
+}
+
+// initializeVMap will build a nested map structure to hold all our version information
+func initializeVMap() map[string]map[string]string {
+	versionMap := make(map[string]map[string]string, types.UnknownType)
+
+	for i := 0; i < int(types.UnknownType); i++ {
+		versionMap[GetResponseTypeURL(types.ResponseType(i))] = make(map[string]string)
+	}
```

Comment:
Is this going to work with custom resource types? Answer can be no, we can do in a follow up unless there are major blockers

Reply:
I don't believe so no. Right now we only build the version map off the pre-defined resource types in the resource package so if a consumer were to use custom resource types I don't think it would work as of right now

Reply:
Sure, let's file an issue to track?